File: Leetcode/46. Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if len(nums) == 1:
            return [nums]
        else:
            arr = []
            for i in range(len(nums)):
                n = nums[i]
                other = nums[:i] + nums[i+1:]
                temp = []
                for j in self.permute(other):
                    arr.append([n]+j)
            return arr

File: Leetcode/1-999/46. Permutations.java
class Solution {
    public void backtrack(List<List<Integer>> ans, int[] nums, List<Integer> temp){
        if(temp.size() == nums.length){
            ans.add(new ArrayList<>(temp));
        }
        else{
            for(int i : nums){
                if(temp.contains(i)) continue;
                temp.add(i);
                backtrack(ans, nums, temp);
                temp.remove(temp.size()-1);
            }
        }
    }
    
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new LinkedList<>();
        backtrack(ans, nums, new LinkedList<>());
        return ans;
    }
}
